File: 169/Majority_Element.c
int majorityElement(int* nums, int numsSize) {
    if (!nums || numsSize == 0) return 0;
    int i, num = nums[0], count = 1;
    for (i = 1; i < numsSize; i++) {
        if (count == 0) {
            num = nums[i];
            count++;
        } else if (num == nums[i]) {
            count++;
        } else {
            count--;
        }
    }
    return num;
}

File: 169/majority_element.py
import math


class Solution:

    def majorityElementBrutalForce(self, nums: list):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return
        threshhold = math.floor(len(nums) / 2.0)
        # cache
        checked = set()
        for i in nums:
            if i in checked:
                continue
            else:
                checked.add(i)
            count = nums.count(i)
            if count > threshhold:
                return i

    def majorityElement(self, nums: list):
        if not nums:
            return
        count = 1
        major = nums[0]
        for i in nums[1:]:
            if i != major:
                count -= 1
                if count == 0:
                    # reset
                    count = 1
                    major = i
            else:
                count += 1

        return major


